make-zip

```diff
@@ -1,80 +1,183 @@
+
+    This function should be updated if the bundle format changes.
+
+    :param kit_version: The kit version the bundle is targeted for.
+    :param wifi_ssid: The SSID for the WiFi access point.
+    :param wifi_psk: The PSK for the WiFi access point.
+    :param wifi_region: Region that the WiFi hotspot will be used in.
+    """
+    return {
+        "bundle": {
+            "version": "2.0.0",
+        },
+        "kit": {
+            "name": "Student Robotics",
+            "version": kit_version,
+        },
+        "wifi": {
+            "ssid": wifi_ssid,
+            "psk": wifi_psk,
+            "enabled": True,
+            "region": wifi_region,
+        },
+    }
+
+
+def create_usercode_bundle(
+    team_tla: str,
+    usercode_dir: Path,
+    output_file: Path,
+    wifi_psk: str,
+    wifi_region: str,
+) -> None:
+    """
+    Create a usercode bundle for use with SR2021+ kit.
+
+    :param team_tla: TLA of the team that the zip is created for.
+    :param usercode_dir: Directory of usercode, containing a robot.py entrypoint.
+    :param output_file: File to write the zip data to, usually robot.zip.
+    :param wifi_psk: WiFi PSK used to secure robot hotspot.
+    :param wifi_region: Region that the WiFi hotspot will be used in.
+    :raise UnableToCreateUsercodeBundleException: An error occurred during creation.
+    """
+    entrypoint_path = usercode_dir / ENTRYPOINT
+
+    # Validate arguments
+    if not usercode_dir.exists() or not usercode_dir.is_dir():
+        raise UnableToCreateUsercodeBundleException(
+            f"Usercode directory ({usercode_dir}) is not a directory or doesn't exist.",
+        )
+
+    if not entrypoint_path.exists():
+        raise UnableToCreateUsercodeBundleException(
+            f"Usercode directory ({usercode_dir}) does not contain a {ENTRYPOINT} file.",
+        )
+
+    if len(team_tla) == 0:
+        raise UnableToCreateUsercodeBundleException("Team TLA cannot be empty.")
+
+    # Create the ZIP
+    with ZipFile(output_file, "w", compression=ZIP_DEFLATED) as zip_file:
+        # Add the usercode
+        for folderName, subfolders, filenames in os.walk(usercode_dir):
+            relativeFolder = os.path.relpath(folderName, usercode_dir)
+            for filename in filenames:
+                relative_path = os.path.join(relativeFolder, filename)
+                real_path = os.path.join(folderName, filename)
+
+                # Check that the file should not be ignored
+                if not is_file_ignored(relative_path):
+                    zip_file.write(real_path, relative_path)
+
+        # Add bundle.toml
+        bundle_data = get_bundle_dict(
+            "0.0.0.0dev",  # This should eventually be fetched from an API
+            f"robot-{team_tla}",
+            wifi_psk,
+            wifi_region,
+        )
+        bundle_str = toml.dumps(bundle_data)
+        zip_file.writestr("bundle.toml", bundle_str)
+
+
+def main() -> None:
```

Comment:
drive-by thought -- I tend to design short scripts to keep the argument parsing in a separate function than the `main`, even when I do also try to push most of the logic into other utils as well. The separation between argument parsing and even using those arguments helps avoid leakage between the two as things grow as well as aiding discoverability in the code. In that regard I tend to structure things a bit like:

``` python
def whatever_util(...):
    ...

def main(args):
    ...

def parse_args():
    ...

if __name__ == '__main__':
    exit(main(parse_args()))
```